sdk/src/main/scala/com/horizen/SidechainState.scala

```diff
@@ -434,6 +448,58 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage,
     getFeePayments(withdrawalEpochNumber, None)
   }
 
+  def getAppStateVersion(versions: util.List[Array[Byte]]) : VersionTag =
+  {
+    val applicationStateVersionsSet = versions.asScala.map{x=>new ByteArrayWrapper(x)}.toSet
+
+    if (applicationStateVersionsSet.size == 1) {
+      log.debug("All application state storages are consistent, common version: " + BytesUtils.toHexString(applicationStateVersionsSet.head))
+      bytesToVersion(applicationStateVersionsSet.head.data())
+    } else {
+      log.warn("Not all application state storages are consistent")
+      versions.asScala.zipWithIndex.foreach{
+        case(x,i) => log.warn(s"app storage ${i} version: ${bytesToVersion(x)}")
+      }
+      // application has internal storages which are not aligned, return a null version tag
+      bytesToVersion(new ByteArrayWrapper(0).data())
+    }
+  }
+
+  // Check that all storages are consistent and in case try some rollbacks.
+  // Return the state and common version, throw an exception if some unrecoverable misalignment has been detected
+  def ensureStorageConsistencyAfterRestore: Try[(SidechainState, ByteArrayWrapper)] = {
+    // updates are in order:
+    //      appState--> utxoMerkleTreeStorage --> stateStorage --> forgerBoxStorage
+    val appStateVersion = getAppStateVersion(applicationState.getStoragesVersionList)
+    val versionFb       = bytesToId(forgerBoxStorage.lastVersionId.get.data())
+
+    val result = if (appStateVersion == versionFb) {
+      // TODO csw capability will be optional, related utxo mrkl tree storage might even be empty
+      if (utxoMerkleTreeStorage.lastVersionId.isDefined) {
+        val versionUmt = bytesToId(utxoMerkleTreeStorage.lastVersionId.get.data())
+        require(appStateVersion == versionUmt)
+      }
+
+      val versionSt  = bytesToId(stateStorage.lastVersionId.get.data())
+      require(appStateVersion == versionSt)
+
+      log.debug("All state storages are consistent")
+      val baw = new ByteArrayWrapper(versionToBytes(version))
+      Success(this, baw)
+    } else {
+      log.debug("state storages are not consistent")
+
+      val rolledBackState = rollbackTo(idToVersion(versionFb))
```

Comment:
The reason is because when we check wallet storage consistency, we already ensured that state and history are consistent, and since wallet  is update in between state and history  (state --> wallet --> history) we can not have any wallet inconsistency and therefore should throw an exception.